Annotation serialisation framework

jakelishman opened the pull request:
This is a roll-up of the current description of the annotation framework.  I'll be back later today to finish it up.  All of the hard work and design work is done, I just have to finish writing a bit more docs, write the QPY bit of the implementation (which will be small, like the OQ3 implementation, because all the complexity is pushed to users), and the tests (which will spill).

<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Implementation list:

- [x] add annotation base class
- [x] add documentation on what annotations are, and what's required for subclassing
- [x] fill in the TODOs in the `Annotation` documentation I just noticed are still there
- [x] add serialisation framework for OpenQASM 3
- [x] implement serialisation to OpenQASM 3
- [x] implement deserialisation from OpenQASM 3 (might be on `qiskit-qasm3-import`, and isn't so urgent)
- [x] add serialisation framework for QPY including documentation of new format
- [x] implement serialisation to QPY
- [x] implement deserialisation from QPY
- [x] sort out one borked QPY test (might be indicative that serialisation/deserialisation doesn't work yet)
- [x] write all the new tests...
- [x] release note

### Details and comments

jakelishman committed:
Add `Annotation` base class and `BoxOp` support

This adds the basics of support for a custom annotation framework to
Qiskit.  An `Annotation` is a piece of custom-defined information (from
Qiskit's perspective; it's most likely to be transpiler-pass defined,
even downstream) that is localised to a particular instruction or scope.
Currently, the only supported location for use of annotations is `box`.

The meaning of the `Annotation` may currently be completely custom.  We
expect in the future to have a coarse set of validity ranges that
annotations associate themselves with, to allow the transpiler to
enforce that annotations are not violated by compiler passes that do not
understand them.  For now, it's up to the user not to call compiler
passes that invalidate annotations.

Annotations are closely related to the OpenQASM 3 concept of the same
name.  Serialisation to and from OpenQASM 3 and QPY is supported by
additional user-defined classes, whose interfaces are defined in
`qiskit.circuit.annotation`.  QPY version 15 is introduced, which makes
it a collaborative serialisation format, with well-defined slots in the
output stream for custom serialisers to store their data, including
state.  OpenQASM 3 annotation serialisers must be stateless, since that
language does not provide (nor intend) the same arbitrary co-operation.
QPY has this additional support for statefulness because we anticipate
larger amounts of data being transferred in annotations in QPY format
for subsequent processing.

alexanderivrii commented on crates/circuit/src/annotation.rs:
```diff
@@ -0,0 +1,104 @@
+///
+/// This is a framework for structuring additional metadata that can be attached to :class:`.BoxOp`
+/// instructions within a :class:`.QuantumCircuit` and :class:`.DAGCircuit` in ways that can be
+/// tracked and consumed by arbitrary tranpsiler passes, including custom passes that are not in
```
typo: "tranpsiler"

kevinhartman commented on qiskit/circuit/annotation.py:
```diff
@@ -0,0 +1,328 @@
+    guarantees, and it is strongly recommended that annotation-serialisation subclasses maintain
+    this.  In particular, it is suggested that any non-trivial serializer includes "version"
+    information for the serializer in its total "state" (see :meth:`dump_state`), and the
+    deserialization should make
```
should make...?!

kevinhartman commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -6785,32 +6788,47 @@ def box(
                 copy=False,
             )
+        if body_or_annotations is ...:
+            annotations = () if annotations is ... else annotations
```
Very minor, but should this be `[]` instead of `()`? Since I think we call `list` on the annotations coming in to the `BoxOp`.

jakelishman replied:
`list` always creates a new list from its input, so it doesn't make much difference, and has a specialised fast path for conversion from tuple.  The empty tuple is a singleton in CPython, so actually this is probably _slightly_ faster because it avoids an intermediate allocation.

jakelishman replied:
also, I probably originally used `()` as the kwarg default (since it's immutable and safe, unlike `[]`), then realised that I needed to be able to tell if the object had been passed or not, and `()` is a valid value

kevinhartman commented on qiskit/circuit/annotation.py:
```diff
@@ -0,0 +1,328 @@
+        This method is the mirror of :meth:`load_annotation`.
+
+        Args:
+            namespace: the namespace that this serializer was accessed under.  This may be an
+                ancestor of the annotation.
```
Just curious, what would the expected use case for this be? Would an implementation decide to return `NotImplemented` in some case based on this, or would we expect this to be part of the payload they write out?

jakelishman replied:
The idea of namespaces having `.`-separated identifiers is just straight from the OQ3 spec, and then I leant into it in the rest of the design.  It's nice, especially in conjunction with the serialisation state in QPY, because you can have a single serialiser object that can handle multiple different _types_ of annotation, and use a shared state for all of them.  It gives you a framework for having multiple `Annotation` subclasses that share some higher logical state (like references to pulse channels or noise models, maybe), so you only want a single serialiser, and the namespace lets you distinguish things.

The `NotImplemented` thing came from when I was intending to have a catch-all serialiser for QPY that looked things up using globally-configured OpenQASM 3 ones, and so the exception is managed as a proper type state during serialisation/deserialisation, which will make it rather cheaper to handle when moving down to Rust.  I didn't want that to be a special case, so I just said it's available at all levels of the lookup - I can imagine a situation where you might have a "specialised" serialiser that handles only some entries in a dotted namespace, and you want a general serialiser for the whole namespace as a backup.

kevinhartman commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -6785,32 +6788,47 @@ def box(
+        if body_or_annotations is ...:
+            annotations = () if annotations is ... else annotations
+        elif annotations is not ...:
+            raise TypeError("QuantumCircuit.box() got multiple values for argument 'annotations'")
```
This might be slightly unclear. Could we mention that annotations were passed both positionally as the first argument and via the 'annotations' kwarg?

jakelishman replied:
This error message is (supposed to be) identical to what you'd get in Python if you did something like
```python
def f(x, y):
   pass
f(2, x=3)
```
which is logically what the function is supposed to look like you're doing.

kevinhartman commented on qiskit/qasm3/exporter.py:
```diff
@@ -1129,6 +1142,15 @@ def build_for_loop(self, instruction: CircuitInstruction) -> ast.ForLoopStatemen
+        """Use the custom serializers to construct an annotation object."""
+        for namespace in iter_namespaces(annotation.namespace):
+            if (serializer := self.annotation_serializers.get(namespace, None)) is not None and (
+                payload := serializer.dump(annotation)
```
Are there any checks we should do on the payload to make sure it's supported by QASM? Disclaimer: I haven't looked at the QASM spec to see what's supported.

jakelishman replied:
It just has to be a single line of valid utf-8 text.  Python should enforce valid utf-8 by nature of the return value being `str`, and I just didn't include the check for `\n` since imo that'd be something that should live in the test suites of the custom serialisers.

kevinhartman commented on qiskit/qpy/__init__.py:
```diff
@@ -398,6 +425,141 @@ def open(*args):
+    struct INSTRUCTION;
+    uint8_t name[name_size];
+    uint8_t label[label_size];
+    uint8_t register[conditional_reg_name_size]; (1)
```
Should this `(1)` be after `struct INSTRUCTION` instead?

jakelishman replied:
I'd intended it to be a note read as part of understanding the line it's actually on - I can say from just having _used_ the note myself during debugging that it at least made sense to me haha.  I can see the argument for it being on the `struct INSTRUCTION` line if you prefer.

kevinhartman commented on qiskit/qpy/binary_io/circuits.py:
```diff
@@ -41,11 +46,81 @@
+            if (serializer := self.potential_serializers.get(namespace, None)) is not None:
+                if (out := serializer.dump_annotation(annotation)) is not NotImplemented:
+                    del self.potential_serializers[namespace]
+                    index = len(self.serializers)
```
Should we have size checking on the number of serializers to make sure we don't exceed the width of a `u32`? I don't think we do that for anything else in QPY though, anyway.

jakelishman replied:
that's like 4 billion or something - I think you'd have larger problems if you were exceeding that

kevinhartman reviewed (commented):
This looks pretty solid to me.

The one concern I have that ties in closely to the control flow oxidation stuff I'm working on is the Rust-side representation of annotations. We're going to need to use a `Py<PyAny>` to represent these, and since you've got them as fields of `BoxOp` as opposed to `params`, they'll end up roughly here, directly on the `ControlFlow` operation enum: `OperationRef::ControlFlow(ControlFlow::Box { annotations, ... })`. At the moment, `ControlFlow` derives `PartialEq`, but with a `Py<PyAny>` inside, we'd have to implement it manually and acquire the GIL (if we want it implemented at all).

If you stick it into `params`, that shouldn't be an issue since we already compare those with the `Python` token. I've got a `ControlFlowView<'a>` enum FWIW that ties together what we store in the operation (`ControlFlow`) and the params into a unified view:

```rust
pub enum ControlFlowView<'a, T> {
    Box(Option<&'a Duration>, &'a T),
    BreakLoop,
    ContinueLoop,
    ForLoop {
        indexset: &'a [usize],
        loop_param: Option<&'a PyObject>,
        body: &'a T,
    },
    IfElse {
        condition: &'a Condition,
        true_body: &'a T,
        false_body: Option<&'a T>,
    },
    Switch {
        target: &'a Target,
        cases_specifier: Vec<(&'a Vec<CaseSpecifier>, &'a T)>,
    },
    While {
        condition: &'a Condition,
        body: &'a T,
    },
}
```

Separately, I'm curious how you envision annotations being exposed to Rust in the future. Any thoughts there?

I'll have another look once you've got the tests and remaining documentation TODOs sorted.

jakelishman commented:
Longer term, I want `annotation` to be a property of `PackedInstruction` not a special behaviour of `box` (I think I wrote that somewhere in the docs), and to be the general way that we do tagging of temporary local inter-pass attributes onto single instructions inside the compiler IR, so various passes don't try (poorly) to reinvent the wheel, and so the pass manager can reason about them in a limited way. That also happens to match how OQ3 handles annotations.  I just stuck to only `Box` for the first iteration, which is where it _needs_ to be used at first.

In Rust, I will probably want to hide the list of annotations in the sort of "extra attributes" indirection we used to have for `condition`/`unit`/`duration`/`label`, and then got reduced down just to `label`, because we expect the use of it to be rare relative to instructions.

As to representation of `Annotation` in Rust itself: my approximate idea is that there will be a pretty limited "properties" API that annotations have to specify how they behave under certain very abstract IR manipulations (valid under co-variant transformations of types X, Y Z; valid under contra-variant transformations; valid only for the verbatim set of instructions at the time of tagging; etc, where the "transformations" are things like "the physical qubits used", "the physical qubit 2q connections used", etc).  The results of those configurations will all be Rust native, accessed by some sort of virtual-dispatch trait method or something.  All the custom data of annotations will be custom and generally invisible; only the passes intended to interpret a given annotation (can use the namespace to determine, perhaps) will know what type to interpret the pointer as, and thus to use it, etc.

I don't have the full Rust implementation fleshed out - it's very similar to the problem about how we'll represent custom gates in a way that isn't Python-specific, so e.g. Rust (including Qiskit itself) can put temporary custom objects into the circuit.  I have rough ideas, but for the time being, I'm just throwing it into the "sort out with the other custom stuff" bucket, since annotations are needed by the primitives teams in the short term.

jakelishman commented:
Ok, that should be the last time I force push to this branch - I'll just do everything additively from now.  The major changes in the last force push were to modify how the QPY serialiser subclasses work - previously they were passing a namespace to all four of the methods, but actually, the way the interface is designed, it was only possible for a single look-up key to be associated with any one _instance_, so half of the passing was irrelevant.

As a note: the current implementation makes a new annotation-state context for each subcircuit; we don't share the annotation state when we enter a `Box`, for example, so annotations in a nested box will be serialised into a clean state.   This is consistent with how the same registers will end up getting re-serialised, etc, within control-flow blocks in QPY.  In a future version of QPY, we should make control flow more natively supported, and have the subcircuits share the same annotation contexts (and register contexts, `Var` contexts, etc), since that'll be more efficient for serialisation.

jakelishman committed:
Complete documentation of subclassing

jakelishman committed:
Fix circuit equality of boxes

jakelishman commented:
Qiskit/qiskit-qasm3-import#50 adds an implementation of annotated boxes, and provides an API that this library can then wrap.  I tested that PR against this branch.

jakelishman committed:
Add support for annotations in `qasm3.load` API

This does a small amount of version negotiation with
`qiskit_qasm3_import`, which in turn checks the version of Qiskit.

jakelishman committed:
Include comment on equality

jakelishman committed:
Add OpenQASM 3 import/export tests

jakelishman committed:
Merge remote-tracking branch 'ibm/main' into box/annotation

jakelishman committed:
Fix previous parameter-expression QPY documentation

ElePT commented on qiskit/qpy/__init__.py:
```diff
@@ -407,11 +431,147 @@ def open(*args):
+and its associated serializer state in the ``ANNOTATION_HEADER``.
+
+
+Changes within PARAM_EXPR_ELEM_V13
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The struct itself is unchanged.  However, for a ``PARAM_EXPR_ELEM_V13`` representing a
+:meth:`.ParameterExpression.subs` call (with ``op_code = 15``, and therefore ``lhs_type = 'p'`` and
+``rhs_type = 'n'``), the trailing :ref:`qpy_mapping` now maps keys of the raw bytes of the
+:class:`.Parameter` UUIDs to the substituted values.  Previously (in QPY versions 13 and 14), this
+mapping stored the parameter names as the keys.
+
```
This looks much better than the original wording. It's also nice to have more changes that justify a QPY version change, instead of a tiny detail

jakelishman committed:
Support `BoxOp` in `GenericBackendV2`

jakelishman committed:
Fix annotation handling in `BoxOp.replace_blocks`

jakelishman committed:
Add transpilation integration test

jakelishman committed:
Fix GenericBackendV2 test

jakelishman committed:
Add `iter_namespaces` test

jakelishman committed:
Add QPY tests

jakelishman committed:
Add release note

abbycross commented on test/python/circuit/test_annotation.py:
```diff
@@ -0,0 +1,30 @@
+from qiskit.circuit import annotation
+from test import QiskitTestCase  # pylint: disable=wrong-import-order
+
+# Most of the tests for this module are elsewhere, such as for OpenQASM 3 serialisation being in the
```
```suggestion
# Most of the tests for this module are elsewhere, such as for OpenQASM 3 serialization being in the
```

jakelishman replied:
Thanks Abby - this particular one is just a code comment, and you're fighting a losing battle trying to localise those (I do my best in public docs, but there's a limit when it's just me "talking" in comments).

abbycross replied:
I hear that! Copyediting is a constant losing battle, so I'm used to that :D Forgive me when I made such edits, but the American spelling per IBM Style guidance is branded in my brain.

abbycross commented on test/python/qasm3/test_export.py:
```diff
@@ -3139,3 +3139,102 @@ def test_old_alias_classical_registers_option(self):
         self.assertEqual(dumps_experimental(qc, allow_aliasing=True), expected_qasm)
+
+    def test_annotations(self):
+        """Test that the annotation-serialisation framework works."""
```
```suggestion
        """Test that the annotation-serialization framework works."""
```

abbycross commented on test/python/qpy/test_circuit_load_from_qpy.py:
```diff
@@ -352,3 +360,230 @@ def __bool__(self):
+        triggered_not_implemented = False
+
+        class Serializer(annotation.QPYSerializer):
+            # The idea with this family of serialisers is that we say the "value" we expect to see
```
```suggestion
            # The idea with this family of serializers is that we say the "value" we expect to see
```

jakelishman commented:
Abby: thanks for the effort with the localisation.  In general, code comments (starting `#` or `//`) don't need localising, and even docstrings in files that start `test/` generally don't either - neither of those are publicly visible at all.

I try my best with the spelling when I'm writing public-facing documentation, but it's a (small) effort for me that I just don't bother with when it's not for publication - most of our test docstrings and code comments are barely even grammatically correct...

If I were you, I wouldn't waste time looking at those.  Comments on the public ones (docstrings in `qiskit/` or `crates/`) are very welcome though.

jakelishman commented on qiskit/qasm3/exporter.py:
```diff
@@ -172,6 +175,10 @@ def __init__(
                 set to the empty string to disable indentation.
             experimental: any experimental features to enable during the export.  See
                 :class:`ExperimentalFeatures` for more details.
+            annotation_serializers: a mapping of namespaces to annotation serializers.  When an
```
```suggestion
            annotation_handlers: a mapping of namespaces to annotation serializers.  When an
```

jakelishman committed:
Fix parameter documentation

kevinhartman commented on test/python/qasm3/test_export.py:
```diff
@@ -3139,3 +3139,102 @@ def test_old_alias_classical_registers_option(self):
+
+        class MyHandler(annotation.OpenQASM3Serializer):
+            def load(self, namespace, payload):
+                return NotImplemented  # Unused in test
```
Should we `assert(false)` here instead?

jakelishman replied:
oh sure, I can swap it to `raise NotImplementedError` - that's a better way of saying what I meant

jakelishman replied:
Done in 572baf0cf5.

kevinhartman reviewed (commented):
This looks good to me, thanks! Just one minor comment in the tests.

> Longer term, I want annotation to be a property of PackedInstruction not a special behaviour of box
> ...
> In Rust, I will probably want to hide the list of annotations in the sort of "extra attributes" indirection we used to have for condition/unit/duration/label, and then got reduced down just to label, because we expect the use of it to be rare relative to instructions.

Ah alright, that sounds good then. I would've imagined the same regarding how we'd store them on `PackedInstruction`.

The other context you've provided on the future of this in Rust is helpful.

jakelishman committed:
Make test logic errors exceptions

kevinhartman reviewed (approved):
LGTM!
